### DISCUSSION TITLE
Refactoring of component dialog box

### DISCUSSION THREAD

**Discussion by iwbnwif** [ra3xdh/qucs_s#1052] [Ideas]
I have refactored the component dialog box here: https://github.com/iwbnwif/qucs_s/tree/new_comp_dialog

The main aim of this change is to switch from editing equations by selecting them in a table and then typing in a separate field, to entering them using free text. At the moment the editing is very free(!) and it might be that it needs to be more constrained.

The main problem I started off trying to solve is when you accidentally overwrite an existing equation by clicking 'apply' instead of 'add'. I think that free text is an easier way of entering equations, especially longer lines that didn't fit completely in the edit box.

Some basic syntax highlighting is included for Ngspice keywords only. It would also be nice to detect some basic mistakes and underline them - but this is a future enhancement.

Any rows that do not include an equals '=' are ignored and will be lost. Maybe in future it would be nice  to include comments?

This is quite a substantial change and although there are a few jobs still to do, I would welcome any feedback at this stage. I can only practically test the changes on Linux at the moment so bug reports from other platforms are welcome.

```
/*
  TODO:
  1. Auto update sweep step / sweep points for log sweeps
  2. Translated text?
  3. Add special property names - i.e., for log sweeps (per decade instead of step) - DONE
  4. Update components from SPICE file.
  5. Implement highlighting - DONE
  6. Have "Export" as a check box, or option list for equations.
*/
```

A second change is to edit component properties directly in the table rather than in a separate dialog. Again, I am sure there are some corner cases that I have missed.

I have constrained my changes to just two files. The rest of the simulation and drawing is unchanged.

**Comment by ra3xdh** [comment]
Please open a PR with this new feature and mark it as WIP. This will allow to test possible merge conflicts and apply CI. Also there exits PR #1049 which changes some class members names. 

In general a new equation editor looks better than old one. The old component dialog was inherited from Qt3 which had no table cells editing feature and also would benefit from redesign. I did a quick test of the new editing system and found the following issues:

* Everything is selected after switching to the *Properties* tab of the simulation component. Normal devices are not affected. 
![image](https://github.com/user-attachments/assets/1b3e1d0c-09eb-4766-b5ca-11dd1f9d6e6b)
* The Qucs equations have a simulation selection combo box. This property has no effect for the equations. It should be shown only for Nutmeg equations. 
![image](https://github.com/user-attachments/assets/5ef26ccb-6f7e-46c3-9a19-ac0635127569)
* The syntax highlighting and advanced equation editing is also needed for .PARAM device. It is also a type of equation, but it is not attached to the simulation. 
* There exists B-type sources and EDD devices that also may have complex equations inside. 
![image](https://github.com/user-attachments/assets/be574532-5ca3-402e-8343-35a09bf4966b)
* The syntax highlighting is also needed for SPICE files editor. 
![image](https://github.com/user-attachments/assets/c0896218-8c57-4ac3-8f7a-5986d79e5fff)

**Reply by zergud** [reply]
> The syntax highlighting is also needed for SPICE files editor

we may use [QScintilla](https://riverbankcomputing.com/software/qscintilla/download) as default text editor and provide custom lexer for spice etc.

**Reply by iwbnwif** [reply]
Yes, that was my thought too but it does add a dependency unless QScintilla is already being used?

**Reply by zergud** [reply]
QScintilla is not used yet...
new dependency is not a problem for me..
but GPLv3 vs GPLv2 incompatibility may be (I'm not good at law)

**Comment by iwbnwif** [comment]
Thank you for these comments. 

> Please open a PR with this new feature and mark it as WIP. This will allow to test possible merge conflicts and apply CI. Also there exits PR https://github.com/ra3xdh/qucs_s/pull/1049 which changes some class members names.

I have opened a PR but unfortunately seem to have messed up my attempt at squashing the old commits :( I will merge the changes in #1049 once it has been accepted.

> Everything is selected after switching to the Properties tab of the simulation component. Normal devices are not affected.

This should be fixed in the latest commit.

> The Qucs equations have a simulation selection combo box. This property has no effect for the equations. It should be shown only for Nutmeg equations.

> The syntax highlighting and advanced equation editing is also needed for .PARAM device. It is also a type of equation, but it is not attached to the simulation.

I will fix these two shortly.

> There exists B-type sources and EDD devices that also may have complex equations inside.

This is more tricky to address as EDD have both multiple properties and multiple equations. I think the real solution to this would be able to open an equation editor from a property table line (similar to Freecad's f(x) button in dimension edit boxes but with a wider editor if you've ever used that). I don't think that I will be able to manage that in this change.

> The syntax highlighting is also needed for SPICE files editor.

Noted, we can pull out the syntax highlighting class into a separate file once complete and reuse it for the text editor. But again, I think that will be separate to this change.

**Comment by iwbnwif** [comment]
PR #1054 opened and this discussion will now continue there.
